File: nodejs-tools/nodejs-compiler-tools/lexer-xml/src/readTextNode.ts
import { EOF_CHARACTER, Lexer } from "@vighnesh153/lexer-core";
import { XmlTokenType } from "./tokens.ts";

const LEFT_ANGLE_BRACKET = "<";

export function readTextNode(lexer: Lexer<XmlTokenType>): string {
  const textNodeBuilder: string[] = [];

  let currCh = lexer.inputReader.currentCharacter;
  while (
    lexer.inputReader.peekCharacter() !== LEFT_ANGLE_BRACKET &&
    currCh !== EOF_CHARACTER
  ) {
    textNodeBuilder.push(currCh);
    lexer.inputReader.readNextCharacter();
    currCh = lexer.inputReader.currentCharacter;
  }

  if (lexer.inputReader.peekCharacter() === LEFT_ANGLE_BRACKET) {
    textNodeBuilder.push(lexer.inputReader.currentCharacter!);
  }

  return textNodeBuilder.join("");
}

File: tools-jvm/XmlFormatter/src/main/kotlin/dev/vighnesh153/xml/formatter/lexer/readTextNode.kt
package dev.vighnesh153.xml.formatter.lexer

internal fun XmlLexer.readTextNode(): String {
    val textNodeBuilder = StringBuilder()

    while (inputReader.currChar.isNotEOF() && inputReader.peekChar() != LEFT_ANGLE_BRACKET) {
        textNodeBuilder.append(inputReader.currChar)
        inputReader.readNextChar()
    }

    if (inputReader.peekChar() == LEFT_ANGLE_BRACKET) {
        textNodeBuilder.append(inputReader.currChar)
    }

    return textNodeBuilder.toString()
}
